lambdas/s3hash/src/t4_lambda_s3hash/__init__.py

```diff
@@ -0,0 +1,388 @@
+from __future__ import annotations
+
+import asyncio
+import base64
+import contextlib
+import contextvars
+import enum
+import functools
+import hashlib
+import math
+import os
+import typing as T
+
+import aiobotocore.config
+import aiobotocore.response
+import aiobotocore.session
+import botocore.exceptions
+import pydantic
+import tenacity
+import types_aiobotocore_s3.type_defs as T_S3TypeDefs
+from types_aiobotocore_s3.client import S3Client
+
+from t4_lambda_shared.utils import get_quilt_logger
+
+logger = get_quilt_logger()
+
+DEFAULT_CONCURRENCY = int(os.environ["MPU_CONCURRENCY"])
+MULTIPART_CHECKSUMS = os.environ["MULTIPART_CHECKSUMS"] == "true"
+SERVICE_BUCKET = os.environ["SERVICE_BUCKET"]
+
+SCRATCH_KEY_SERVICE = "user-requests/checksum-upload-tmp"
+SCRATCH_KEY_PER_BUCKET = ".quilt/.checksum-upload-tmp"
+
+
+S3: contextvars.ContextVar[S3Client] = contextvars.ContextVar("s3")
+
+
+NonEmptyStr = pydantic.constr(min_length=1, strip_whitespace=True)
+
+
+class AWSCredentials(pydantic.BaseModel):
+    key: NonEmptyStr
+    secret: NonEmptyStr
+    token: NonEmptyStr
+
+
+# Isolated for test-ability.
+def get_user_boto_session(
+    credentials: AWSCredentials,
+) -> aiobotocore.session.AioSession:
+    s = aiobotocore.session.get_session()
+    s.set_credentials(credentials.key, credentials.secret, credentials.token)
+    return s
+
+
+@contextlib.asynccontextmanager
+async def aio_context(credentials: AWSCredentials, concurrency: pydantic.PositiveInt):
+    session = get_user_boto_session(credentials)
+    config = aiobotocore.config.AioConfig(max_pool_connections=concurrency)
```

Comment:
I think you want to limit the number of concurrent requests, but IIUC this just limits number of connections **kept** in the pool i.e. you still can get more than `max_pool_connections` concurrent request, so I think you'd want to use [`semaphore`](https://docs.python.org/3/library/asyncio-sync.html#asyncio.Semaphore) somewhere.

Reply:
well, limiting max_pool_connections fixes "dropouts" (reduces their number to something acceptable at least), so...

> use [semaphore](https://docs.python.org/3/library/asyncio-sync.html#asyncio.Semaphore) somewhere

probably this is a good idea

Reply:
> well, limiting max_pool_connections fixes "dropouts" (reduces their number to something acceptable at least), so...

you're not limiting it but increasing it from default 10
https://botocore.amazonaws.com/v1/documentation/api/latest/reference/config.html
that can reduce "dropouts", because it needs to open less new connections so there is less overhead

btw what kind of errors did you get?

Reply:
> you're not limiting it but increasing it from default 10

i mean i was experimenting with that limit

> btw what kind of errors did you get?

i don't remember rn, but most likely hitting some local connection limits, not AWS rate-limiting or smth